### DISCUSSION TITLE
Restrict heatmap to an arbitrary area

### DISCUSSION THREAD

**Discussion by vadmbertr** [maplibre/maplibre-gl-js#125] [Q&A]
Hi!
Thanks a lot for this fork.

I'm using the heatmap layer and I would like to restrict the coloring to an arbitrary area (let say a polygon shape). To make it clearer: inisde the polygon coloring would follow the heatmap-color property of the layer and outside the polygon I'd like to have a transparent or greyed color. I also attached a picture reflecting it (note that the actual heatmap I have inside the polygon is quite different).
Is there any way I could do it?

Vadim

![heatmap](https://user-images.githubusercontent.com/36510417/112818037-61bfb000-9083-11eb-957c-bc714020bf89.png)

**Comment by nickpeihl** [comment]
This appears to be a user question, and I think we'd like to direct questions to the [Discussions](https://github.com/maplibre/maplibre-gl-js/discussions) section. Would you be willing to post this there instead? This allows us to use issues for verified bug reports and feature requests.

If you want to post this as a feature request, I think we might need to know more about the use case. Why is this necessary and how might others benefit from it?

**Reply by vadmbertr** [reply]
Yeah sure.

But at the same time indeed I have the feeling that it could be an interesting feature to have:
 - You have sensors (or data) for a particular area (not the whole globe) and you want to visualize one of your indicators using a heatmap. It's quite missleading I think to have the heatmap extends to the entire globe simply because you suggest a measurement where in fact you have no data and therefore no idea of what that measurement would be.
 - In my case, I'm estimating the density of vehicles along roads and one of the dispay is at the vehicle level. I plot both the vehicles as circles and using those points I add a heatmap layer. It's that layer that goes to "infinity".
 - Having a way to restrict the layer to a particular area (a polygon seems appropriate) would improve a lot the understanding of the visualization in my opinion.

If you feel that it's too restricted to a particular use case then I will move my question to the Discussions section!

**Reply by nickpeihl** [reply]
Apologies. I was hasty and already moved this to Discussions. IMO, the heat map algorithm used by maplibre-gl may be too naive for visualizing your use case. An area with extremely dense road networks (such as Manhattan) might have road densities clobbering each other, even with the restrictions you propose. An algorithm that uses network analysis might be more appropriate, but maybe out of scope for this project. 

But I'm only one person expressing an opinion. If you feel strongly about this requirement, feel free to copy and paste your expanded use case into a new issue as a feature request.

**Reply by vadmbertr** [reply]
That's fine, no worries!
Indeed the implementation leaves little room for customization and my use case is a bit at its limits. On the other hand having road densities clobbering each other is not an issue since the aim is to provide a 2D visualization instead of a 1D along the roads.

I came up with a hackish solution for my use case so I don't need to open a feature request and this thread is well placed in the Discussions section. In case other people are interested in having more details about my workaround I will be happy to share it.
